File: src/test/scala/akkastudy/simpleactor/scala/SimpleActorTest.scala
package akkastudy.simpleactor.scala

import org.scalatest.FlatSpec
import akka.actor.{Props, ActorSystem}

class SimpleActorTest extends FlatSpec {
  behavior of "A simple actor"

  it should "receive our message in Scala" in {
    val system = ActorSystem("MySystem")
    val myActor = system.actorOf(Props[SimpleActorScala], name = "simpleActorScala")
    myActor ! "Simple Test"
    myActor ! "test"
  }

  it should "receive our message in Scala using a factory" in {
    val system = ActorSystem("MySystem")
    val myActor = system.actorOf(Props(new SimpleActorScala), name = "simpleActorScala")
    myActor ! "Simple Test"
    myActor ! "test"
  }

  it should "be at location akka://MySystem/user/simpleActorJava" in {
    val system = ActorSystem("MySystem")
    system.actorOf(Props[SimpleActorScala], name = "simpleActorJava")
    val actorSelection = system.actorSelection("akka://MySystem/user/simpleActorJava")
    actorSelection ! "Simple Test"
  }

  it should "throw send a dead letter if the actor is not found" in {
    val system = ActorSystem("MySystem")
    system.actorOf(Props[SimpleActorScala], name = "simpleActorJava")
    val actorSelection = system.actorSelection("akka://MySystem/user/somethingElseIShouldn\'tBeLookingFor")
    actorSelection ! "Simple Test"
  }
}

File: src/test/java/akkastudy/simpleactor/java/SimpleActorTest.java
package akkastudy.simpleactor.java;

import akka.actor.ActorRef;
import akka.actor.ActorSelection;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.japi.Creator;
import org.junit.Test;
import scala.concurrent.Await;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

public class SimpleActorTest {

    @Test
    public void testActor() throws Exception {
        ActorSystem system = ActorSystem.create("MySystem");
        ActorRef myActor = system.actorOf(
                Props.create(SimpleActorJava.class),
                "simpleActorJava");


        System.out.println(Thread.currentThread().getName());
        myActor.tell("Bueno!", system.deadLetters());
        Thread.sleep(10000);
        Await.result(system.terminate(), Duration.apply(10, TimeUnit.SECONDS));

    }

    @Test
    public void testActorWithAFactory() throws Exception {
        ActorSystem system = ActorSystem.create("MySystem");
        ActorRef myActor = system.actorOf(
                Props.create(SimpleActorJava.class,
                        new Creator<SimpleActorJava>() {
                    @Override
                    public SimpleActorJava create() throws Exception {
                        return new SimpleActorJava();
                    }
                }), "simpleActorJava"
        );
        myActor.tell("Bueno!", system.deadLetters());
        Thread.sleep(10000);
        Await.result(system.terminate(), Duration.apply(10, TimeUnit.SECONDS));
    }

    @Test
    public void testActorSelection() throws Exception {
        ActorSystem system = ActorSystem.create("MySystem");
        system.actorOf(
                Props.create(SimpleActorJava.class),
                "simpleActorJava");

        Thread.sleep(2000);

        System.out.println
                (Thread.currentThread().getName());

        ActorSelection myActor =
                system.actorSelection
                        ("akka://MySystem/user/simpleActorJava");

        myActor.tell("Bueno!", system.deadLetters());
        Thread.sleep(4000);
        Await.result(system.terminate(), Duration.apply(10, TimeUnit.SECONDS));
    }
}
